[CI] ManyShardsIT testRejection failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #4969 / ubuntu-1804_platform-support-unix](https://gradle-enterprise.elastic.co/s/v7wt54jrge5bk)
- [elasticsearch-periodic-platform-support #4969 / rhel-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/tvfmzom67o4g4)
- [elasticsearch-periodic-platform-support #4956 / rhel-9_platform-support-unix](https://gradle-enterprise.elastic.co/s/hx4iw4hbrghy6)
- [elasticsearch-periodic-platform-support #4917 / oraclelinux-7_platform-support-unix](https://gradle-enterprise.elastic.co/s/cq5vd645btz2k)
- [elasticsearch-periodic-platform-support #4917 / oraclelinux-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/7h6flqtbttdng)
- [elasticsearch-periodic-platform-support #4912 / rocky-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/3l72rlxudvv7e)
- [elasticsearch-periodic #4932 / single-processor-node-tests](https://gradle-enterprise.elastic.co/s/2ufvfjk3g7rb4)
- [elasticsearch-periodic-platform-support #4897 / rhel-7_platform-support-unix](https://gradle-enterprise.elastic.co/s/xtngntjlhf6mc)
- [elasticsearch-periodic-platform-support #4887 / almalinux-8-aarch64_checkpart3_platform-support-arm](https://gradle-enterprise.elastic.co/s/uqo3qnhjh5gio)

**Reproduction Line:**
```
./gradlew ':x-pack:plugin:esql:internalClusterTest' --tests "org.elasticsearch.xpack.esql.action.ManyShardsIT.testRejection" -Dtests.seed=26D51210A7272351 -Dtests.locale=sr-Latn-RS -Dtests.timezone=Asia/Irkutsk -Druntime.java=22
```

**Applicable branches:**
8.15

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.esql.action.ManyShardsIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testRejection),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: Leftover exchanges ExchangeService{sinks=[M6axhOYHQBqTPJ2FDslNUA:1343]} on node node_s1
```

**Issue Reasons:**
- [8.15] 9 failures in test testRejection (1.8% fail rate in 488 executions)
- [8.15] 6 failures in pipeline elasticsearch-periodic-platform-support (37.5% fail rate in 16 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/ManyShardsIT.java
+++ b/x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/ManyShardsIT.java
@@ -14,9 +14,13 @@ import org.elasticsearch.action.bulk.BulkRequestBuilder;
 import org.elasticsearch.action.index.IndexRequest;
 import org.elasticsearch.action.support.WriteRequest;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
 import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
 import org.elasticsearch.compute.operator.exchange.ExchangeService;
+import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.rest.RestStatus;
 import org.elasticsearch.search.MockSearchService;
@@ -26,6 +30,7 @@ import org.elasticsearch.transport.RemoteTransportException;
 import org.elasticsearch.transport.TransportChannel;
 import org.elasticsearch.transport.TransportResponse;
 import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.esql.plugin.ComputeService;
 import org.elasticsearch.xpack.esql.plugin.QueryPragmas;
 import org.hamcrest.Matchers;
 import org.junit.Before;
@@ -54,6 +59,18 @@ public class ManyShardsIT extends AbstractEsqlIntegTestCase {
         plugins.add(MockSearchService.TestPlugin.class);
         plugins.add(MockTransportService.TestPlugin.class);
         return plugins;
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return CollectionUtils.appendToCopy(super.nodePlugins(), InternalExchangePlugin.class);
+    }
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal, Settings otherSettings) {
+        return Settings.builder()
+            .put(ExchangeService.INACTIVE_SINKS_INTERVAL_SETTING, TimeValue.timeValueMillis(between(3000, 5000)))
+            .build();
     }
 
     @Before
@@ -113,32 +130,64 @@ public class ManyShardsIT extends AbstractEsqlIntegTestCase {
     }
 
     public void testRejection() throws Exception {
-        String[] nodes = internalCluster().getNodeNames();
-        for (String node : nodes) {
-            MockTransportService ts = (MockTransportService) internalCluster().getInstance(TransportService.class, node);
-            ts.addRequestHandlingBehavior(ExchangeService.EXCHANGE_ACTION_NAME, (handler, request, channel, task) -> {
-                handler.messageReceived(request, new TransportChannel() {
-                    @Override
-                    public String getProfileName() {
-                        return channel.getProfileName();
-                    }
+        DiscoveryNode dataNode = randomFrom(internalCluster().clusterService().state().nodes().getDataNodes().values());
+        String indexName = "single-node-index";
+        client().admin()
+            .indices()
+            .prepareCreate(indexName)
+            .setSettings(
+                Settings.builder()
+                    .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                    .put("index.routing.allocation.require._name", dataNode.getName())
+            )
+            .setMapping("user", "type=keyword", "tags", "type=keyword")
+            .get();
+        client().prepareIndex(indexName)
+            .setSource("user", "u1", "tags", "lucene")
+            .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+            .get();
 
-                    @Override
-                    public void sendResponse(TransportResponse response) {
-                        channel.sendResponse(new RemoteTransportException("simulated", new EsRejectedExecutionException("test queue")));
-                    }
+        MockTransportService ts = (MockTransportService) internalCluster().getInstance(TransportService.class, dataNode.getName());
+        CountDownLatch dataNodeRequestLatch = new CountDownLatch(1);
+        ts.addRequestHandlingBehavior(ComputeService.DATA_ACTION_NAME, (handler, request, channel, task) -> {
+            handler.messageReceived(request, channel, task);
+            dataNodeRequestLatch.countDown();
+        });
 
-                    @Override
-                    public void sendResponse(Exception exception) {
-                        channel.sendResponse(exception);
-                    }
-                }, task);
+        ts.addRequestHandlingBehavior(ExchangeService.EXCHANGE_ACTION_NAME, (handler, request, channel, task) -> {
+            ts.getThreadPool().generic().execute(new AbstractRunnable() {
+                @Override
+                public void onFailure(Exception e) {
+                    channel.sendResponse(e);
+                }
+
+                @Override
+                protected void doRun() throws Exception {
+                    assertTrue(dataNodeRequestLatch.await(30, TimeUnit.SECONDS));
+                    handler.messageReceived(request, new TransportChannel() {
+                        @Override
+                        public String getProfileName() {
+                            return channel.getProfileName();
+                        }
+
+                        @Override
+                        public void sendResponse(TransportResponse response) {
+                            channel.sendResponse(new RemoteTransportException("simulated", new EsRejectedExecutionException("test queue")));
+                        }
+
+                        @Override
+                        public void sendResponse(Exception exception) {
+                            channel.sendResponse(exception);
+                        }
+                    }, task);
+                }
             });
-        }
+        });
+
         try {
             AtomicReference<Exception> failure = new AtomicReference<>();
             EsqlQueryRequest request = new EsqlQueryRequest();
-            request.query("from test-* | stats count(user) by tags");
+            request.query("from single-node-index | stats count(user) by tags");
             request.acceptedPragmaRisks(true);
             request.pragmas(randomPragmas());
             CountDownLatch queryLatch = new CountDownLatch(1);
@@ -151,9 +200,7 @@ public class ManyShardsIT extends AbstractEsqlIntegTestCase {
             assertThat(ExceptionsHelper.status(failure.get()), equalTo(RestStatus.TOO_MANY_REQUESTS));
             assertThat(failure.get().getMessage(), equalTo("test queue"));
         } finally {
-            for (String node : nodes) {
-                ((MockTransportService) internalCluster().getInstance(TransportService.class, node)).clearAllRules();
-            }
+            ts.clearAllRules();
         }
     }
 

